### PR TITLE
Converting Tensorflow models into PyTorch 

### PR DESCRIPTION
**Pull request by SamChou05**

### CODE REVIEW

File: afqinsight/nn/pt_models.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+torch, has_torch, _ = optional_package("torch", trip_msg=torch_msg)
+if has_torch:
+    import torch.nn as nn
+else:
+    # Since all model building functions start with Input, we make Input the
+    # tripwire instance for cases where pytorch is not installed.
+    Input = TripWire(torch_msg)
+    print("test")
+
+class mlp4_torch(nn.Module):
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: Naming: no need to add `_torch` to class names here, because users will know where they are importing from. In fact, might be beneficial to use the exact same API here as in the tensorflow module. 

And that means that all the public interfaces here are functions (not classes!) that take inputs and give a model ready to fit as outputs. Or does that not make sense for Pytorch?

Comment (arokem) [reply afqinsight/nn/pt_models.py]: One solution for this might be something like: 

```

class MLP4(...): 

   ... 

```

Followed by: 

```
def mlp4(...): 

   mymlp4 = MLP4(...)
   ... 

   return mymlp4
``

File: afqinsight/nn/pt_models.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        super(mlp4_torch, self).__init__()
+        self.flatten = nn.flatten()
+        self.dropout_1 = nn.Dropout(0.1)
+        self.dropout_2 = nn.Dropout(0.2)
+        self.dropout_3 = nn.Dropout(0.3)
+
+        self.l0 = nn.Linear(input_shape, 500)
+        self.l1 = nn.ReLU(500,500)
+        self.l2 = nn.ReLU(500,500)
+        self.l3 = nn.ReLU(500,500)
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: I think there needs to be a `Linear` before every `ReLu` to create the fully connected model.

File: afqinsight/nn/pt_models.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        x = self.l2(x)
+        x = self.dropout_2(x)
+        x = self.l3(x)
+        x = self.dropout_3(x)
+        x = self.l4(x)
+        x = self.output(x)
+
+        return x
+    
+class cnn_lenet_torch(nn.Module):
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: Same comment as above regarding naming and design.

File: afqinsight/nn/pt_models.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                )
+            )
+            conv_layers.append(nn.ReLU())
+            conv_layers.append(nn.MaxPool1d(kernel_size=2))
+            
+        self.conv_layers = nn.Sequential(*conv_layers)
+        
+        self.flatten = nn.flatten()
+        self.fc1 = nn.ReLU(input_shape, 120)
+        self.fc2 = nn.ReLU(120, 84)
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: Fully connected implies => `Linear`. The `ReLu` is the activation function (non-linearity) that follows the fully connected step, so this can't be right (I think?).

File: afqinsight/nn/pt_models.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+       
+        x = self.fc1(x)
+        x = self.dropout_1(x)
+        x = self.fc2(x)
+        x = self.dropout_2(x)
+        x = self.fc3(x)  
+        x = self.output(x)
+        return x
+
+class cnn_vgg_torch(nn.Module):
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: Same comment about naming and design.

File: afqinsight/nn/pt_models.py (hunk 2 above)
Comment (arokem) [reply afqinsight/nn/pt_models.py]: I think you probably want to use `nn.Sequential` to chain together layers into models.

File: afqinsight/nn/pt_models.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        
+        #idea is to create a list that sequential can use to
+            #create the model with the order of layers?
+        conv_layers = []
+
+        for i in range(self.n_conv_layers):
+            conv_layers.append(
+                nn.Conv1d(
+                    # what is in channels?
+                    in_channels= input_shape[i],
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: My understanding is that `in_channels` for the first conv layer is the number of channels in the input data. That is, `input_shape[1]`. The `out_channel` is the number of conv filters. Then, in every subsequent layer, `in_channel` would be the number of filters in the previous layer?

**Review by arokem — COMMENTED** [review]
A few comments to get you started

File: afqinsight/nn/pt_models.py [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+torch, has_torch, _ = optional_package("torch", trip_msg=torch_msg)
+if has_torch:
+    import torch.nn as nn
+else:
+    # Since all model building functions start with Input, we make Input the
+    # tripwire instance for cases where pytorch is not installed.
+    Input = TripWire(torch_msg)
+    print("test")
+
+class mlp4(nn.Module):
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: Per PEP8, classes should be CapsCase and functions should be lower_case (with snake_case wherever relevant).

File: afqinsight/nn/pt_models.py [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+        
+        #idea is to create a list that sequential can use to
+            #create the model with the order of layers?
+        conv_layers = []
+
+        for i in range(self.n_conv_layers): 
+            if i == 0:
+                conv_layers.append(
+                    nn.Conv1d(
+                        # what is in channels?
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: Can we get rid of this comment?

File: afqinsight/nn/pt_models.py [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+            self.output_activation = None
+        
+
+    def forward(self, x):
+        x = self.model(x)
+        if self.output_activation:
+            x = self.output_activation(x)
+        return x
+
+    def cnn_lenet(input_shape, n_classes):
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: Maybe this bit of boilerplate -- for each class, you implement a function that initializes an instance of this class -- could be automated? Have a function that takes class names as inputs and generates these functions as its outputs?

File: afqinsight/nn/pt_models.py [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+
+def lstm2(input_shape, n_classes):
+    lstm2 = lstm2(input_shape, n_classes, output_activation = torch.softmax, verbose=False)
+    return lstm2
+    
+class blstm1(nn.Module):
+    def __init__(self, input_shape, n_classes, output_activation = torch.softmax, verbose=False):
+        super(blstm1, self).__init__()
+        self.model = nn.Sequential(
+            nn.LSTM(input_shape[1], 100, bidirectional=True),
```
Comment (arokem) [inline afqinsight/nn/pt_models.py:311]: I think this is identical (or very similar) to blstm2? I think blstm1 is a lot simpler.

File: afqinsight/nn/tf_models.py [hunk 11]
```diff
@@ -30,20 +29,34 @@
         concatenate,
     )
     from tensorflow.keras.models import Model
+    print("woo")
```
Comment (arokem) [inline afqinsight/nn/tf_models.py]:
```suggestion
```

File: afqinsight/nn/tf_models.py [hunk 12]
```diff
@@ -30,20 +29,34 @@
...
 else:
     # Since all model building functions start with Input, we make Input the
     # tripwire instance for cases where tensorflow is not installed.
     Input = TripWire(keras_msg)
+    print("test")
```
Comment (arokem) [inline afqinsight/nn/tf_models.py]:
```suggestion
```

File: afqinsight/__init__.py [hunk 13]
```diff
@@ -6,4 +6,4 @@
 from .datasets import *  # noqa
 from .pipeline import *  # noqa
 from .transform import *  # noqa
-from ._version import version as __version__  # noqa
+# from ._version import version as __version__  # noqa
```
Comment (arokem) [inline afqinsight/__init__.py]:
```suggestion
from ._version import version as __version__  # noqa
```

File: afqinsight/nn/tests/test_compare_py_tf.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+    lstm1_pt,
+    lstm1v0_pt,
+    lstm2_pt,
+    mlp4_pt,
+)
+from afqinsight.nn.tf_models import cnn_lenet, cnn_vgg, lstm1, lstm1v0, lstm2, mlp4
+from afqinsight.nn.utils import compare_models
+
+
+# test for mlp4
```
Comment (arokem) [inline afqinsight/nn/tests/test_compare_py_tf.py]: The function name already says this, so maybe we don't need these comments 😄

File: afqinsight/nn/tests/test_compare_py_tf.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+
+# test for LSTM2
+def test_lstm2():
+    pytorch_lstm2 = lstm2_pt(input_shape=(784, 1), n_classes=10)
+    tensorflow_lstm2 = lstm2(input_shape=(784, 1), n_classes=10, verbose=True)
+    assert compare_models(pytorch_lstm2, tensorflow_lstm2)
+
+
+# def test_blstm1():
```
Comment (arokem) [inline afqinsight/nn/tests/test_compare_py_tf.py]: Uncomment and figure out... eventually 😄

File: afqinsight/tests/test_nn.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
+import torch
```
Comment (arokem) [inline afqinsight/tests/test_nn.py]: This file needs to be removed too

File: afqinsight/nn/tests/test_compare_py_tf.py [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+    assert compare_models(pytorch_blstm1, tensorflow_blstm1)
+
+
+def test_blstm2():
+    pytorch_blstm2 = blstm2_pt(input_shape=(784, 1), n_classes=10)
+    tensorflow_blstm2 = blstm2(input_shape=(784, 1), n_classes=10, verbose=True)
+    assert compare_models(pytorch_blstm2, tensorflow_blstm2)
+
+
+# def test_lstm_fcn():
```
Comment (arokem) [inline afqinsight/nn/tests/test_compare_py_tf.py:77]:
```suggestion
# The following tests are commented out because comparing models 
# with both FCN and LSTM components is very fiddly, but we have 
# independently verified that these models have identical structure 
# We are keeping these tests around, in case someone wants to come back 
# here and fix this up.
# def test_lstm_fcn():
```

File: afqinsight/nn/tf_models.py [hunk 18]
```diff
@@ -244,6 +250,8 @@ def blstm2(input_shape, n_classes, output_activation="softmax", verbose=False):
     return model
 
 
+# FCN IS FULLY CONVULUTIONAL NETWORK
+# LSTM IS LONG SHORT TERM MEMORY
```
Comment (arokem) [inline afqinsight/nn/tf_models.py]: I think we can remove these comments

File: afqinsight/nn/pt_models.py [hunk 19]
```diff
@@ -0,0 +1,526 @@
...
+
+    def forward(self, x):
+        x = self.model(x)
+        if self.output_activation:
+            x = self.output_activation(x)
+        return x
+
+
+def cnn_lenet_pt(input_shape, n_classes):
+    cnn_lenet_Model = CNN_LENET(
```
Comment (arokem) [inline afqinsight/nn/pt_models.py:128]:
```suggestion
    """ 
    Lenet convolutional network. 
    
    Parameters 
    ----------
    input_shape : tuple 
        The shape of the input. 
    n_classes: int
        The number of classes (for classification tasks) or 
        1 (for regression tasks).
        
    Returns 
    -------
    CNN_LENET class instance.
    """
    cnn_lenet_Model = CNN_LENET(
```

Comment (arokem) [reply afqinsight/nn/pt_models.py:128]: This follows the numpy docstring standard: https://numpydoc.readthedocs.io/en/latest/format.html

File: afqinsight/nn/tf_models.py [hunk 20]
```diff
@@ -244,6 +246,8 @@ def blstm2(input_shape, n_classes, output_activation="softmax", verbose=False):
...
```
Comment (arokem) [inline afqinsight/nn/tf_models.py]:
```suggestion
```

**Comment by arokem** [comment]
Congratulations on your first merged PR @SamChou05! 🎉
